api/search.go

```diff
@@ -0,0 +1,374 @@
+	// Status is the results data for this file for each relevant run.
+	Status []SearchRunResult `json:"status"`
+}
+
+// SearchResponse contains a response to search API calls, including specific
+// runs whose results were searched and the search results themselves.
+type SearchResponse struct {
+	// Runs is the specific runs for which results were retrieved. Each run, in
+	// order, corresponds to a Status entry in each SearchResult in Results.
+	Runs []shared.TestRun `json:"runs"`
+	// Results is the collection of test results, grouped by test file name.
+	Results []SearchResult `json:"results"`
+}
+
+// summary is the golang type for the JSON format in pass/total summary files.
+type summary map[string][]int
+
+type byName []SearchResult
+
+func (r byName) Len() int           { return len(r) }
+func (r byName) Swap(i, j int)      { r[i], r[j] = r[j], r[i] }
+func (r byName) Less(i, j int) bool { return r[i].Name < r[j].Name }
+
+type readable interface {
+	Get(string) ([]byte, error)
+}
+
+type readWritable interface {
+	readable
+	Put(string, []byte) error
+}
+
+type httpReadable struct {
+	ctx context.Context
+}
+
+func (hr httpReadable) Get(url string) ([]byte, error) {
+	client := urlfetch.Client(hr.ctx)
+	r, err := client.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	defer r.Body.Close()
+
+	if r.StatusCode != 200 {
+		return nil, fmt.Errorf("Unexpected status code from %s: %d", url, r.StatusCode)
+	}
+
+	var data []byte
+	if data, err = ioutil.ReadAll(r.Body); err != nil {
+		return nil, err
+	}
+
+	return data, nil
+}
+
+type gzipReadWritable struct {
+	delegate readWritable
+}
+
+func (gz gzipReadWritable) Get(key string) ([]byte, error) {
+	zipped, err := gz.delegate.Get(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return gunzipData(zipped)
+}
+
+func gunzipData(zipped []byte) ([]byte, error) {
+	gzReader, err := gzip.NewReader(bytes.NewReader(zipped))
+	if err != nil {
+		return nil, err
+	}
+	defer gzReader.Close()
+
+	return ioutil.ReadAll(gzReader)
+}
+
+func (gz gzipReadWritable) Put(key string, unzipped []byte) error {
+	var buf bytes.Buffer
+	gzipData(unzipped, &buf)
+	return gz.delegate.Put(key, buf.Bytes())
+}
+
+func gzipData(unzipped []byte, zipped *bytes.Buffer) error {
+	bufWriter := bufio.NewWriter(zipped)
+	defer bufWriter.Flush()
+	gzWriter := gzip.NewWriter(bufWriter)
+	defer gzWriter.Close()
+
+	_, err := gzWriter.Write(unzipped)
+	return err
+}
+
+type memcacheReadWritable struct {
+	ctx context.Context
+}
+
+func (mc memcacheReadWritable) Get(key string) ([]byte, error) {
+	item, err := memcache.Get(mc.ctx, key)
+	if item == nil {
+		return nil, err
+	}
+
+	return item.Value, err
+}
+
+func (mc memcacheReadWritable) Put(key string, value []byte) error {
+	log.Printf("Writing %d-bytes to memcache object: %s", len(value), key)
+	return memcache.Set(mc.ctx, &memcache.Item{
+		Key:        key,
+		Value:      value,
+		Expiration: 48 * time.Hour,
+	})
+}
+
+type sharedImpl interface {
```

Comment:
Based on offline discussion, this is mostly for easier testing, so I'm fine with it for now although ideally we should really change the `shared` package to provide an interface and an impl.

However, I still can't come to terms with an interface named impl... Perhaps `sharedInterface`, `abstractShared`?